music21/osmd/osmd.py

```diff
@@ -0,0 +1,222 @@
+                urllib.request.urlretrieve(self.script_url, self.osmd_file)
+
+            # since we can't link to files from a notebook (security risk) inject into file.
+            script_content = open(self.osmd_file).read()
+            script = script.replace('"{{offline_script}}"', json.dumps(script_content), 1)
+
+        else:
+            script = script.replace('"{{script_url}}"', json.dumps(self.script_url), 1)
+        return script
+
+    @staticmethod
+    def addDefaultPartName(score):
+        '''
+        Due to a bug in OpenSheetMusicDisplay if a <part-name> is not provided it will default
+        to an auto-generated random string. This method is used to fix that by default.
+
+
+        >>> import music21
+        >>> # display two scores to demonstrate difference.
+        >>> s = music21.converter.parse("tinyNotation: 3/4 E4 r f# g=lastG trip{b-8 a g} c4~ c")
+        >>> _ = s.show('osmd',fixPartName=False)
+        >>> music21.osmd.ConverterOpenSheetMusicDisplay.addDefaultPartName(s)
+        >>> _ = s.show('osmd',fixPartName=False)
+        '''
+        # If no partName is present in the first instrument, OSMD will display the ugly 'partId'
+        allInstruments = list(score.getInstruments(returnDefault=False, recurse=True))
+
+        if not allInstruments:
+            defaultInstrument = Piano()
+            defaultInstrument.instrumentName = ' '
+            score.insert(0.0, defaultInstrument)
+        elif not allInstruments[0].instrumentName:
+            # instrumentName must not be '' or None
+            allInstruments[0].instrumentName = ' '
+
+
+class TestExternal(unittest.TestCase):
+
+    @unittest.skipUnless(hasInstalledIPython, "skipping since IPython not installed")
+    def testOpenSheetMusicDisplayRuns(self):
+        from music21 import corpus
+
+        s = corpus.parse('bwv66.6')
+        ConverterOpenSheetMusicDisplay().show(s, None)
+
+    @unittest.skipUnless(hasInstalledIPython, "skipping since IPython not installed")
+    def testOpenSheetMusicDisplayColors(self):
+        s = music21.converter.parse("tinyNotation: 3/4 E4 r f# g=lastG trip{b-8 a g} c4~ c")
+
+        for n in s.recurse().getElementsByClass('Note'):
+            n.style.color = 'red'
+            if str(n.pitch) == 'G4':
+                n.style.color = 'blue'
+        for n in s.recurse().getElementsByClass('Rest'):
+            n.style.color = 'green'
+
+        s.show('osmd')
+
+class Test(unittest.TestCase):
+
+    def testAddsDefaultPartId(self):
+        from music21 import converter
+        s = converter.parse("tinyNotation: 3/4 E4 r f#")
+
+        ConverterOpenSheetMusicDisplay.addDefaultPartName(s)
+        firstInstrumentObject = s.getInstruments(returnDefault=True, recurse=True)[0]
+        self.assertNotEqual(firstInstrumentObject.instrumentName, None)
+        self.assertNotEqual(firstInstrumentObject.instrumentName, '')
+
+
+if __name__ == "__main__":
+    import music21
+
+    music21.mainTest(TestExternal)
```

Comment:
Myke will chime in if I get this wrong, but usually the command line call here runs the `Test` class instead of `TestExternal`

Reply:
That's usually true, but here I'm okay with making it a list of [Test, TestExternal] because the main thing someone would want to check is "did the notation get generated?"  I think we do that in `graph` (or we should) and maybe flat.io